discovery_child_development/notebooks/02_explore_concepts.py

```diff
@@ -11,164 +13,302 @@
+    concepts_cleaned_df.groupby("display_name").size().reset_index(name="n_works")
+)
+unique_concepts_df.head()
 
+# %%
+# Create embeddings for the unique concepts
 t0 = time()
-unique_concepts_384 = model.encode(unique_concepts, show_progress_bar=True)
+unique_concepts_384 = model.encode(
+    unique_concepts_df["display_name"].tolist(), show_progress_bar=True
+)
 print(f"vectorization done in {time() - t0:.3f} s")
 
-umap_vectors = umap.UMAP(
-    n_neighbors=15, n_components=50, random_state=42
-).fit_transform(unique_concepts_384)
-
-concept_clusters = hdbscan.HDBSCAN(
-    min_cluster_size=10,  # min_samples=15,
-    metric="euclidean",
-    cluster_selection_method="eom",
-    prediction_data=True,
-).fit(umap_vectors)
-
-
-umap_2d = umap.UMAP(n_neighbors=15, n_components=2, random_state=42).fit_transform(
-    unique_concepts_384
+# %%
+# Define HDBSCAN search parameters
+hdbscan_search_params = {
+    "min_cluster_size": [10, 20, 50],
+    "min_samples": [1, 10, 15],
+    "cluster_selection_method": ["leaf"],
+    "prediction_data": [True],
+}
+
+umap_params = {
+    "n_components": 50,  # apparently hdbscan does not work very well with more than 50 components
+    "n_neighbors": 10,
+    "min_dist": 0.5,
+    "spread": 0.5,
+}
+
+# %%
+# reduce dimensionality of the embeddings
+unique_concepts_50 = cau.umap_reducer(
+    unique_concepts_384, umap_params, random_umap_state=SEED
 )
 
-# +
-unique_concepts_df = pd.DataFrame(concepts_df["display_name"].unique())
-unique_concepts_df.columns = ["concept"]
-unique_concepts_df = unique_concepts_df.assign(
-    cluster=concept_clusters.labels_, x=umap_2d[:, 0], y=umap_2d[:, 1]
+# %%
+# %%time
+# Parameter grid search using HDBSCAN
+hdbscan_search_results = hdbscan_param_grid_search(
+    vectors=unique_concepts_50,
+    search_params=hdbscan_search_params,
+    have_noise_labels=False,  # we want everything to be forced into a cluster
 )
 
-unique_concepts_df["opacity"] = unique_concepts_df["cluster"].apply(
-    lambda x: 0.3 if x == -1 else 1.0
+# %%
+hdbscan_search_results
+
+# %%
+# Find HDBSCAN model params with highest silhouette score
+optimal_hdbscan_params = highest_silhouette_model_params(hdbscan_search_results)
+optimal_hdbscan_params
+
+# %% [markdown]
+# We will check Kmeans results as well in case this method performs better than HDBSCAN.
+
+# %%
+# Define K-Means search parameters
+kmeans_search_params = {
+    "n_clusters": [5, 10, 15, 20, 25, 30, 35],
+    "init": ["k-means++"],
+}
+
+# %%
+# %%time
+# Parameter grid search using K-Means
+kmeans_search_results = kmeans_param_grid_search(
+    vectors=unique_concepts_50, search_params=kmeans_search_params, random_seeds=[SEED]
 )
-unique_concepts_df["color"] = unique_concepts_df["cluster"].apply(
-    lambda x: "noise" if x == -1 else x
+
+# %%
+# The silhouette scores for HDBSCAN turn out to be better
+kmeans_search_results
+
+# %%
+# proceed with HDBSCAN clustering, seeing as that leads to a better silhouette score
```

Comment:
Looks like the score was highest for the most granular clustering - which probably makes sense!

Reply:
Yes, I'm now wondering whether using the silhouette score in this way was meaningful or not? I realise now that the smaller your minimum cluster size is, the higher the silhouette score you can get

Reply:
I think it depends on the data distribution. The score probably works well on well-separated clusters (eg like in [this example](https://www.kdnuggets.com/2019/10/clustering-metrics-better-elbow-method.html))

Reply:
Thanks, that is a nice article